File: python/sparc/batch.py
"""
Multi-sample batch processing.
"""

import csv
import subprocess
from concurrent.futures import ProcessPoolExecutor, as_completed
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional, Union


@dataclass
class SampleConfig:
    """Configuration for a single sample."""
    name: str
    r1: Path
    r2: Path
    whitelist: Path
    protocol: str = "10x-3prime-v3"


@dataclass
class BatchResult:
    """Result from batch processing."""
    sample: str
    success: bool
    message: str = ""
    output_dir: Optional[Path] = None


@dataclass
class BatchSummary:
    """Summary of batch processing results."""
    total: int = 0
    succeeded: int = 0
    failed: int = 0
    results: list[BatchResult] = field(default_factory=list)


def parse_manifest(path: Union[str, Path]) -> list[SampleConfig]:
    """
    Parse a sample manifest CSV file.

    Expected columns: sample_name, r1, r2, whitelist[, protocol]

    Parameters
    ----------
    path : str or Path
        Path to manifest CSV file

    Returns
    -------
    list of SampleConfig
        Parsed sample configurations
    """
    samples = []
    with open(path) as f:
        reader = csv.DictReader(f)
        for row in reader:
            name = row.get("sample_name", row.get("name", ""))
            if not name:
                continue
            samples.append(SampleConfig(
                name=name.strip(),
                r1=Path(row["r1"].strip()),
                r2=Path(row["r2"].strip()),
                whitelist=Path(row["whitelist"].strip()),
                protocol=row.get("protocol", "10x-3prime-v3").strip(),
            ))
    return samples


def _run_sample(
    sample: SampleConfig,
    reference: Path,
    output_dir: Path,
    aligner: str = "star",
    max_mismatch: int = 1,
) -> BatchResult:
    """Run the SPARC pipeline on a single sample."""
    sample_output = output_dir / sample.name
    try:
        cmd = [
            "sparc", "pipeline",
            "-1", str(sample.r1),
            "-2", str(sample.r2),
            "-r", str(reference),
            "-o", str(sample_output),
            "-w", str(sample.whitelist),
            "-p", sample.protocol,
            "--aligner", aligner,
            "--max-mismatch", str(max_mismatch),
            "-s", sample.name,
        ]
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        return BatchResult(
            sample=sample.name,
            success=True,
            message="Pipeline completed successfully",
            output_dir=sample_output,
        )
    except subprocess.CalledProcessError as e:
        return BatchResult(
            sample=sample.name,
            success=False,
            message=f"Pipeline failed: {e.stderr}",
        )
    except FileNotFoundError:
        return BatchResult(
            sample=sample.name,
            success=False,
            message="sparc CLI not found in PATH",
        )


class BatchProcessor:
    """
    Process multiple samples from a manifest file.

    Parameters
    ----------
    manifest : str or Path
        Path to sample manifest CSV
    reference : str or Path
        Path to reference genome directory
    output : str or Path
        Output directory
    parallel : int
        Number of samples to process in parallel (default: 1)
    aligner : str
        Aligner to use (default: "star")
    max_mismatch : int
        Maximum barcode mismatch (default: 1)
    """

    def __init__(
        self,
        manifest: Union[str, Path],
        reference: Union[str, Path],
        output: Union[str, Path],
        parallel: int = 1,
        aligner: str = "star",
        max_mismatch: int = 1,
    ):
        self.manifest = Path(manifest)
        self.reference = Path(reference)
        self.output = Path(output)
        self.parallel = parallel
        self.aligner = aligner
        self.max_mismatch = max_mismatch

    def run(self) -> BatchSummary:
        """
        Run the batch processing pipeline.

        Returns
        -------
        BatchSummary
            Summary of all sample results
        """
        samples = parse_manifest(self.manifest)
        self.output.mkdir(parents=True, exist_ok=True)

        summary = BatchSummary(total=len(samples))

        if self.parallel > 1:
            with ProcessPoolExecutor(max_workers=self.parallel) as executor:
                futures = {
                    executor.submit(
                        _run_sample, sample, self.reference, self.output,
                        self.aligner, self.max_mismatch
                    ): sample
                    for sample in samples
                }
                for future in as_completed(futures):
                    result = future.result()
                    summary.results.append(result)
                    if result.success:
                        summary.succeeded += 1
                    else:
                        summary.failed += 1
        else:
            for sample in samples:
                result = _run_sample(
                    sample, self.reference, self.output,
                    self.aligner, self.max_mismatch
                )
                summary.results.append(result)
                if result.success:
                    summary.succeeded += 1
                else:
                    summary.failed += 1

        return summary

File: crates/sparc-cli/src/commands/batch.rs
//! Multi-sample batch processing

use anyhow::{Context, Result};
use clap::Args;
use rayon::prelude::*;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::PathBuf;

#[derive(Args)]
pub struct BatchArgs {
    /// Sample manifest CSV (columns: sample_name,r1,r2,whitelist)
    #[arg(short, long)]
    manifest: PathBuf,

    /// Reference genome directory
    #[arg(short = 'r', long)]
    reference: PathBuf,

    /// Output directory
    #[arg(short, long)]
    output: PathBuf,

    /// Protocol
    #[arg(short, long, default_value = "10x-3prime-v3")]
    protocol: String,

    /// Aligner
    #[arg(long, default_value = "star")]
    aligner: String,

    /// Maximum barcode mismatch
    #[arg(long, default_value = "1")]
    max_mismatch: u32,

    /// Parallel samples
    #[arg(long, default_value = "1")]
    parallel_samples: usize,
}

#[derive(Debug, Clone)]
struct SampleConfig {
    name: String,
    r1: PathBuf,
    r2: PathBuf,
    whitelist: PathBuf,
}

fn parse_manifest(path: &PathBuf) -> Result<Vec<SampleConfig>> {
    let file = File::open(path).context("Failed to open manifest")?;
    let reader = BufReader::new(file);
    let mut samples = Vec::new();

    for (i, line) in reader.lines().enumerate() {
        let line = line?;
        let line = line.trim();

        if i == 0 && line.to_lowercase().contains("sample") {
            continue;
        }
        if line.is_empty() || line.starts_with('#') {
            continue;
        }

        let parts: Vec<&str> = line.split(',').collect();
        if parts.len() < 4 {
            anyhow::bail!(
                "Line {}: expected 4 columns (sample_name,r1,r2,whitelist), got {}",
                i + 1,
                parts.len()
            );
        }

        samples.push(SampleConfig {
            name: parts[0].trim().to_string(),
            r1: PathBuf::from(parts[1].trim()),
            r2: PathBuf::from(parts[2].trim()),
            whitelist: PathBuf::from(parts[3].trim()),
        });
    }

    Ok(samples)
}

pub fn run(args: BatchArgs) -> Result<()> {
    println!("=== SPARC Batch Processing ===\n");

    let samples = parse_manifest(&args.manifest)?;
    println!("Found {} samples in manifest\n", samples.len());

    std::fs::create_dir_all(&args.output)?;

    let results: Vec<(String, Result<()>)> = if args.parallel_samples > 1 {
        let pool = rayon::ThreadPoolBuilder::new()
            .num_threads(args.parallel_samples)
            .build()
            .context("Failed to build thread pool")?;

        pool.install(|| {
            samples
                .par_iter()
                .map(|sample| {
                    let result = process_sample(sample, &args);
                    (sample.name.clone(), result)
                })
                .collect()
        })
    } else {
        samples
            .iter()
            .map(|sample| {
                println!("Processing sample: {}", sample.name);
                let result = process_sample(sample, &args);
                (sample.name.clone(), result)
            })
            .collect()
    };

    println!("\n=== Batch Summary ===");
    let mut succeeded = 0;
    let mut failed = 0;

    for (name, result) in &results {
        match result {
            Ok(()) => {
                println!("  [OK]   {}", name);
                succeeded += 1;
            }
            Err(e) => {
                println!("  [FAIL] {}: {}", name, e);
                failed += 1;
            }
        }
    }

    println!("\nTotal: {} succeeded, {} failed", succeeded, failed);

    if failed > 0 {
        anyhow::bail!("{} samples failed", failed);
    }

    Ok(())
}

fn process_sample(sample: &SampleConfig, args: &BatchArgs) -> Result<()> {
    let sample_output = args.output.join(&sample.name);

    let pipeline_args = super::pipeline::PipelineArgs {
        r1: sample.r1.clone(),
        r2: sample.r2.clone(),
        reference: args.reference.clone(),
        output: sample_output,
        whitelist: sample.whitelist.clone(),
        protocol: args.protocol.clone(),
        sample: sample.name.clone(),
        aligner: args.aligner.clone(),
        max_mismatch: args.max_mismatch,
        min_barcode_qual: 10,
        min_mapq: 30,
        expect_cells: None,
        force_cells: None,
        skip_align: false,
        bam: None,
        min_genes: 200,
        max_genes: 10000,
    };

    super::pipeline::run(pipeline_args)
}
